### PR TITLE
libopae: free MMIO mappings on close

### PR DESCRIPTION
**Pull request by tswhison**
Parameterize wsid_cleanup() with an optional callback so that work can
be done on each mapping when the list is being destroyed. Add call to
destroy handle::mmio_root in fpgaClose().

### CODE REVIEW

**Comment by tswhison** [comment]
@shivarao please check to see if you guys want this in 0.14.

**Comment by shivarao** [comment]
@tswhison we're not putting further changes on 0.14 except if they help with some critical issues we have.

Is this a critical bug fix? What's the impact of not having this? Memory leak?

File: libopae/src/wsid_list.c [hunk 1]
```diff
@@ -116,17 +116,21 @@ bool wsid_del(struct wsid_map **root, uint64_t wsid)
...
-void wsid_cleanup(struct wsid_map **root)
+void wsid_cleanup(struct wsid_map **root, void (*clean)(struct wsid_map * ))
 {
 	if (!*root)
 		return;
 
 	while ((*root)->next) {
 		struct wsid_map *tmp = *root;
 		*root = (*root)->next;
+		if (clean)
```
Comment (michael-adler) [inline libopae/src/wsid_list.c]: Aren't you asking for trouble using multi-line without brackets?

**Comment by tswhison** [comment]
@shivarao Yes, leaked resources and DOS for subsequent processes attempting to obtain the same resource.

File: libopae/src/wsid_list.c (hunk 1 above)
Comment (tswhison) [reply libopae/src/wsid_list.c]: I guess I'm not seeing it.  Isn't clean(tmp); a single statement?

**Comment by shivarao** [comment]
@tswhison  this sounds like a good bug fix but also like an OK bug for our 1.1 Alpha - especially considering no one has reported an issue here as far as we know.

I say leave it off release/0.14 and we'll pick it up when we branch for 1.1 Beta. Let me know if you think my assessment is incorrect

**Comment by michael-adler** [comment]
I think that's a fine choice. It's only a meaningful leak if SW closes an FPGA connection without exiting, which is relatively rare.

**Comment by rpSTTD** [comment]
Is there any existing test test this new logic?

**Comment by tswhison** [comment]
@rpSTTD So, to test this, I commented out the call to fpgaUnmapMMIO() at the end of hello_fpga and stepped through the fpgaClose() call in the debugger, verifying that the callback was invoked to unmap the MMIO region.
As far as introducing an automated test, no.  I can't imagine what that might look like without muddying up the code with some validation hooks.  Do you have any ideas how to test it sans hooks?

**Comment by tswhison** [comment]
@luebbers what do you think about adding a debug assert that handle::mmio_root is NULL at the end of fpgaClose()?  Also, should we apply the corresponding buffer unmap/free logic to a non-empty handle::wsid_root in fpgaClose()?

**Comment by rpSTTD** [comment]
The manual test works for checkin the patch.   https://jira01.devtools.intel.com/browse/OPAE-691 is opened to track automated test development for this patch.

**Pull request merged**